File: test/examples/assignment_teams_sat/assignment_teams_sat.cc































#include <stdlib.h>

#include <numeric>
#include <vector>

#include "absl/strings/str_format.h"
#include "ortools/base/logging.h"
#include "ortools/sat/cp_model.h"
#include "ortools/sat/cp_model.pb.h"
#include "ortools/sat/cp_model_solver.h"



namespace operations_research {
namespace sat {

void AssignmentTeamsSat() {
  

  

  const std::vector<std::vector<int>> costs = {{
      {{90, 76, 75, 70}},
      {{35, 85, 55, 65}},
      {{125, 95, 90, 105}},
      {{45, 110, 95, 115}},
      {{60, 105, 80, 75}},
      {{45, 65, 110, 95}},
  }};
  const int num_workers = static_cast<int>(costs.size());
  std::vector<int> all_workers(num_workers);
  std::iota(all_workers.begin(), all_workers.end(), 0);

  const int num_tasks = static_cast<int>(costs[0].size());
  std::vector<int> all_tasks(num_tasks);
  std::iota(all_tasks.begin(), all_tasks.end(), 0);

  const std::vector<int> team1 = {{0, 2, 4}};
  const std::vector<int> team2 = {{1, 3, 5}};
  

  const int team_max = 2;
  


  

  

  CpModelBuilder cp_model;
  


  

  

  

  

  std::vector<std::vector<BoolVar>> x(num_workers,
                                      std::vector<BoolVar>(num_tasks));
  for (int worker : all_workers) {
    for (int task : all_tasks) {
      x[worker][task] = cp_model.NewBoolVar().WithName(
          absl::StrFormat("x[%d,%d]", worker, task));
    }
  }
  


  

  

  

  for (int worker : all_workers) {
    cp_model.AddAtMostOne(x[worker]);
  }
  

  for (int task : all_tasks) {
    std::vector<BoolVar> tasks;
    for (int worker : all_workers) {
      tasks.push_back(x[worker][task]);
    }
    cp_model.AddExactlyOne(tasks);
  }

  

  LinearExpr team1_tasks;
  for (int worker : team1) {
    for (int task : all_tasks) {
      team1_tasks += x[worker][task];
    }
  }
  cp_model.AddLessOrEqual(team1_tasks, team_max);

  LinearExpr team2_tasks;
  for (int worker : team2) {
    for (int task : all_tasks) {
      team2_tasks += x[worker][task];
    }
  }
  cp_model.AddLessOrEqual(team2_tasks, team_max);
  


  

  

  LinearExpr total_cost;
  for (int worker : all_workers) {
    for (int task : all_tasks) {
      total_cost += x[worker][task] * costs[worker][task];
    }
  }
  cp_model.Minimize(total_cost);
  


  

  

  const CpSolverResponse response = Solve(cp_model.Build());
  


  

  

  if (response.status() == CpSolverStatus::INFEASIBLE) {
    LOG(FATAL) << "No solution found.";
  }
  LOG(INFO) << "Total cost: " << response.objective_value();
  LOG(INFO);
  for (int worker : all_workers) {
    for (int task : all_tasks) {
      if (SolutionBooleanValue(response, x[worker][task])) {
        LOG(INFO) << "Worker " << worker << " assigned to task " << task
                  << ".  Cost: " << costs[worker][task];
      }
    }
  }
  

}
}  

}  


int main(int argc, char** argv) {
  operations_research::sat::AssignmentTeamsSat();
  return EXIT_SUCCESS;
}

File: test/ok/assignment_teams_sat.ts
import { operations_research as op } from '../../src'

function test()
{
    let costs = [
        [90, 76, 75, 70],
        [35, 85, 55, 65],
        [125, 95, 90, 105],
        [45, 110, 95, 115],
        [60, 105, 80, 75],
        [45, 65, 110, 95],
    ];
    let num_workers = costs.length;
    let all_workers = Array.from({ length: num_workers }, (_, i) => i);
    let num_tasks = costs[0].length;
    let all_tasks = Array.from({ length: num_tasks }, (_, i) => i);
    let team1 = [0, 2, 4];
    let team2 = [1, 3, 5];
    let team_max = 2;
    let cp_model = new op.sat.CpModelBuilder();
    let x = Array.from({ length: num_workers }, (_, i) => Array.from(
        { length: num_tasks }, (_, j) => cp_model.NewBoolVar().WithName(
            `x[${i},${j}]`
        )));
    for (let worker of all_workers)
    {
        cp_model.AddAtMostOne(x[worker]);
    }
    for (let task of all_tasks)
    {
        let tasks = [];
        for (let worker of all_workers)
        {
            tasks.push(x[worker][task]);
        }
        cp_model.AddExactlyOne(tasks);
    }
    let team1_tasks = new op.sat.LinearExpr();
    for (let worker of team1)
    {
        for (let task of all_tasks)
        {
            team1_tasks.operator_plus_eq(x[worker][task])
        }
    }
    cp_model.AddLessOrEqual(team1_tasks, team_max);
    let team2_tasks = new op.sat.LinearExpr();
    for (let worker of team2)
    {
        for (let task of all_tasks)
        {
            team2_tasks.operator_plus_eq(x[worker][task])
        }
    }
    cp_model.AddLessOrEqual(team2_tasks, team_max);
    let total_cost = new op.sat.LinearExpr();
    for (let worker of all_workers)
    {
        for (let task of all_tasks)
        {
            total_cost.operator_plus_eq(op.sat.operator_times(x[worker][task], costs[worker][task]))
        }
    }
    cp_model.Minimize(total_cost);
    const response = op.sat.Solve(cp_model.Build());
    if (response.status() === op.sat.CpSolverStatus.INFEASIBLE)
    {
        console.log("No solution found.");
    }
    console.log(`Total cost: ${response.objective_value()}`);
    console.log();
    for (let worker of all_workers)
    {
        for (let task of all_tasks)
        {
            if (op.sat.SolutionBooleanValue(response, x[worker][task]))
            {
                console.log(`Worker ${worker} assigned to task ${task}.  Cost: ${costs[worker][task]}`);
            }
        }
    }
}

test();
